Mathlib/Tactic/Linter/Lint.lean

```diff
@@ -269,4 +269,70 @@ initialize addLinter longLineLinter
+scoped to a single declaration. A non-scoped `open Classical` can hide that some theorem statements
+would be better stated with explicit decidability statements.
+-/
+register_option linter.openClassical : Bool := {
+  defValue := true
+  descr := "enable the openClassical linter"
+}
+
+namespace OpenClassical
+
+/-- Gets the value of the `linter.openClassical` option. -/
+def getLinterHash (o : Options) : Bool := Linter.getLinterValue linter.openClassical o
+
+/-- If `stx` is syntax describing an `open` command, `extractOpenNames stx`
+returns an array of for the opened names with their corresponding syntax
+(omitting any renamed or hidden items). -/
+def extractOpenNames : Syntax → Option (Array (Syntax × Name))
+  | `(command|open $openHiding:openHiding) =>
+    -- The first argument is the identifier we care about.
+    let arg := openHiding.raw[0]
+    some #[(arg, arg.getId)]
+  | `(command|open $openRenaming:openRenaming) =>
+    -- The first argument is the identifier we care about.
+    let arg := openRenaming.raw[0]
+    some #[(arg, arg.getId)]
+  | `(command|open $openOnly:openOnly) =>
+    -- The first argument is the identifier we care about.
+    some #[(openOnly.raw[0], openOnly.raw[0].getId)]
+  | `(command|open $openDecl:openSimple) =>
+    -- The first and only argument of `openDecl` is an array containing all the names we
+    -- are interested in.
+    let namesSyntax := ((openDecl.raw.getArgs).get! 0).getArgs
+    some (namesSyntax.map (fun arg ↦ (arg, arg.getId)))
+  | `(command|open $openScoped:openScoped) =>
+    -- The first argument of `openScoped` is "scoped",
+    -- the second one is an array containing all the names we are interested in.
+    let namesSyntax := (openScoped.raw[1]).getArgs
+    some (namesSyntax.map fun arg ↦ (arg, arg.getId))
+  | _ => none
+
+@[inherit_doc Mathlib.Linter.linter.openClassical]
+def openClassicalLinter : Linter where run := withSetOptionIn fun stx ↦ do
+    unless getLinterHash (← getOptions) do
+      return
+    if (← MonadState.get).messages.hasErrors then
+     return
+    -- TODO: once mathlib's Lean version includes leanprover/lean4#4741, make this configurable
+    unless #[`Mathlib, `test, `Archive, `Counterexamples].contains (← getMainModule).getRoot do
+      return
+    -- If `stx` describes an `open` command, extract the list of opened namespaces.
+    if let some namesSyntaxes := extractOpenNames stx then
+      if let some (stxN, _name) := namesSyntaxes.find? (fun (_, name) ↦ name == `Classical) then
+        Linter.logLint linter.openClassical stxN "\
+        please avoid 'open (scoped) Classical' statements: this can hide theorem statements\n\
+        which would be better stated with explicit decidability statements.\n\
+        Instead, use `open Classical in` for definitions or instances, the `classical` tactic \
+        for proofs.\nFor theorem statements, \
+        either add missing decidability assumptions or use `open Classical in`."
```

Comment:
```suggestion
    for stxN in (extractOpenNames stx).filter (·.getId == `Classical) do
      Linter.logLint linter.openClassical stxN "\
      please avoid 'open (scoped) Classical' statements: this can hide theorem statements\n\
      which would be better stated with explicit decidability statements.\n\
      Instead, use `open Classical in` for definitions or instances, the `classical` tactic \
      for proofs.\nFor theorem statements, \
      either add missing decidability assumptions or use `open Classical in`."
```
then this fixes the parsing!